rep-2010.rst

```diff
@@ -235,6 +313,37 @@ Rationale
 
 TODO
 
+Distributing Type Description via Services, as ``msg`` / ``idl``, on ``rcl``
+----------------------------------------------------------------------------
+
+Sending from Node
+~~~~~~~~~~~~~~~~~
+
+The resolution of the ROS graph and subscriptions afford information about what nodes are publishing or subscribing to any message topic.
+Furthermore, in order for the nodes to register their publisher or subscription in the first place, they must have access to some version of the corresponding message type locally, even if they might have differing versions of the message type.
+Notably, this guarantees that the node itself will have access to the type description, which removes any need to search for the correct service server when querying for type descriptions—a querying node may simply query the node that is advertising the topic.
+
+Type Description as Combination Service Response
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The contents of the type description response sent by the service server should include information that supports most of the major use cases of this feature (namely, tools and type transformation to allow incompatible nodes to communicate.)
+These use cases have orthogonal interests, with the former requiring human-readable descriptions, and the latter preferring machine-usable descriptions.
+
+Furthermore, the constraint to have the type description be usable across middlewares and serialization libraries makes it especially important to send at least whatever information is necessary to regenerate the type support for any middleware or serialization library that is used on the receiving end.
+In this case, because the type support pipeline is a lossy process, and there is a need to ensure that enough information is sent to completely reproduce the original msg file from the msg, losslessly, it makes sense to just send the original ``idl`` or ``msg`` file.
+This also makes it important to send sufficient metadata about the serialization library and version used, since the type support pipeline for the serialized message might have changed, and that support information will be useful to resolve the changes.
+
+Finally, while there could be an argument for sending a losslessly compressed version of the message file, the expected low frequency of queries to the type description service incurs a negligible overhead that heavily reduces the benefit.
```

Comment:
```suggestion
Combining the Raw and Parsed Type Description in the Service Response
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The contents of the "get type description" service response should include information that supports both aforementioned use cases (i.e. tools and runtime type introspection).
These use cases have orthogonal interests, with the former requiring human-readable descriptions, and the latter preferring machine-readable descriptions.

Furthermore, the type description should be useful even across middlewares and serialization libraries and that makes it especially important to send at least the original inputs to the "type support pipeline" (i.e. the process of taking user-defined types and generating all supporting code).
In this case, because the "type support pipeline" is a lossy process, there is a need to ensure that enough information is sent to completely reproduce the original definition of the type, and therefore it makes sense to just send the original ``idl`` or ``msg`` file.

At the same time, it is useful to send information with the original description that makes it easier to process data at the receiving end, as it is often not trivial to get to the "parsed" version of the type description from the original text description.

Finally, while there could be an argument for sending a losslessly compressed version of the message file, the expected low frequency of queries to the type description service incurs a negligible overhead that heavily reduces the benefit.
```